source/common/network/connection_impl.cc

```diff
@@ -927,9 +927,13 @@ ClientConnectionImpl::ClientConnectionImpl(
   if (transport_options) {
     for (const auto& object : transport_options->downstreamSharedFilterStateObjects()) {
       // This does not throw as all objects are distinctly named and the stream info is empty.
-      stream_info_.filterState()->setData(object.name_, object.data_, object.state_type_,
-                                          StreamInfo::FilterState::LifeSpan::Connection,
-                                          object.stream_sharing_);
+      stream_info_.filterState()->setData(
```

Comment:
Could you please extend the current internal_upstream_integration_test.cc 

https://github.com/envoyproxy/envoy/blob/57f4a09ea7dc1ddc56bd9463a53ed9a09d336805/test/extensions/transport_sockets/internal_upstream/internal_upstream_integration_test.cc#L118

 to add integration tests to test :SharedWithUpstreamConnectionOnce. Like maybe having another filter like this but use SharedWithUpstreamConnectionOnce?

https://github.com/envoyproxy/envoy/blob/57f4a09ea7dc1ddc56bd9463a53ed9a09d336805/test/integration/filters/header_to_filter_state.cc#L32

Reply:
Good idea, updated the integration test to cover both cases.

Reply:
Thanks for adding the test to cover "SharedWithUpstreamConnectionOnce" behaves the same as "SharedWithUpstreamConnection" in the case of 1->2.  

I am looking for a test to cover the filter state is not available in 3) in the case of chained loopback configuration, like 1->2->3 with "SharedWithUpstreamConnectionOnce".   Does such test already exist in your tests but I missed?

Reply:
Oh, that doesn't exist. Will need to add a setup with multiple internal listeners.

Reply:
Thanks for holding me to higher standards. Found a bug in the implementation as I wrote the negative integration test.